runtime/tr.source/trj9/z/env/J9CPU.cpp

```diff
@@ -205,33 +206,17 @@ CPU::TO_PORTLIB_get390zLinuxMachineType()
    const int PROC_HEADER_SIZE = sizeof(procHeader) - 1;
 
 #if defined(J9ZTPF)
-   TR_S390MachineType ret_machine = TR_Z10;  /* return value, z/TPF default */
    struct pi1dt *pid;
-   /* machine hardware name */
+   /*
+    * Use the model number from the in-core processor ID table, accessed
+    * via CINFC_CMMPID as the machine type.
+    * The z/TPF OS has strict requirements on supported z/Architecture hardware.
+    * As a result, the model number will necessarily match a valid TR_S390MachineType.
+    * Note - S390MachineTypes, declared above as a static const integer, contains the
+    * hex representation of the (decimal) model number found in the processors ID table.
+    */
    pid = (struct pi1dt *)cinfc_fast(CINFC_CMMPID);
-   int machine = (int)(pid->pi1pids.pi1mslr.pi1mod);
-
-   // Scan list of unsupported machines - We do not initialize the JIT for such hardware.
-   for (int i = 0; i < sizeof(S390UnsupportedMachineTypes) / sizeof(int); ++i)
-      {
-      if (machine == S390UnsupportedMachineTypes[i])
-         {
-         PORT_ACCESS_FROM_ENV(jitConfig->javaVM);
-         j9nls_printf(jitConfig->javaVM->portLibrary, J9NLS_ERROR, J9NLS_J9JIT_390_UNSUPPORTED_HARDWARE, machine);
-         TR_ASSERT(0,"Hardware is not supported.");
-         throw TR::CompilationException();
-         }
-      }
-
-   // Scan list of supported machines.
-   for (int i = 0; i < sizeof(S390MachineTypes) / sizeof(int); ++i)
-      {
-      if (machine == S390MachineTypes[i])
-         {
-         ret_machine = (TR_S390MachineType)machine;
-         }
-      }
-
+   TR_S390MachineType ret_machine = (TR_S390MachineType)(pid->pi1pids.pi1mslr.pi1mod);
```

Comment:
Are you certain `pid->pi1pids.pi1mslr.pi1mod` will always contain a valid processor ID? What happens if this value is not contained within the list of enumerations in `TR_S390MachineType`? Then the value of `ret_machine` is technically undefined. Will this break things?

Reply:
While we get the data at ipl time from the store cpuid instruction (STIDP)  there is no reason to think that this area in main storage couldn't be corrupted at any time.  I will reset the code to do what z/os essentially does; get the character representation of the model, convert it to integer via atoi and then test the integrity of the result.

Reply:
Or simply iterate through the list of machines we support like the code did previously:

```
-   // Scan list of supported machines.
-   for (int i = 0; i < sizeof(S390MachineTypes) / sizeof(int); ++i)
-      {
-      if (machine == S390MachineTypes[i])
-         {
-         ret_machine = (TR_S390MachineType)machine;
-         }
-      }
```